plugin/plugins/filesystem/nodes/pinfs.go

```diff
@@ -0,0 +1,148 @@
+	pd := &PinFS{IPFSBase: newIPFSBase(ctx, "/pinfs", core, ops...)}
+	pd.Qid.Type = p9.TypeDir
+	pd.meta.Mode, pd.metaMask.Mode = p9.ModeDirectory|IRXA, true
+
+	// set up our subsystem, used to relay walk names to IPFS
+	subOpts := []nodeopts.AttachOption{
+		nodeopts.Parent(pd),
+		nodeopts.Logger(logging.Logger("IPFS")),
+	}
+
+	subsystem, err := IPFSAttacher(ctx, core, subOpts...).Attach()
+	if err != nil {
+		panic(err)
+	}
+
+	pd.proxy = subsystem.(walkRef)
+
+	return pd
+}
+
+func (pd *PinFS) Fork() (walkRef, error) {
+	newFid := &PinFS{IPFSBase: pd.IPFSBase.clone()} // root has no paths to walk; don't set node up for change
+	// set new operations context
+	err := newFid.newOperations()
+	return newFid, err
+}
+
+func (pd *PinFS) Attach() (p9.File, error) {
+	pd.Logger.Debugf("Attach")
+
+	newFid := &PinFS{IPFSBase: pd.IPFSBase.clone()} // root has no paths to walk; don't set node up for change
+	// set new fs context
+	err := newFid.newFilesystem()
+	return newFid, err
+}
+
+// PinFS forks the IPFS root that was set during construction
+// and calls step on it rather than itself
+func (pd *PinFS) Step(name string) (walkRef, error) {
+	newFid, err := pd.proxy.Fork()
+	if err != nil {
+		return nil, err
+	}
+	return newFid.Step(name)
+}
+
+func (pd *PinFS) Walk(names []string) ([]p9.QID, p9.File, error) {
+	pd.Logger.Debugf("Walk names %v", names)
+	pd.Logger.Debugf("Walk myself: %v", pd.Qid)
+
+	return walker(pd, names)
+}
+
+func (pd *PinFS) Open(mode p9.OpenFlags) (p9.QID, uint32, error) {
+	pd.Logger.Debugf("Open")
+
+	qid := *pd.Qid
+
+	// IPFS core representation
+	pins, err := pd.core.Pin().Ls(pd.operationsCtx, coreoptions.Pin.Type.Recursive())
+	if err != nil {
+		return qid, 0, err
+	}
+
+	// 9P representation
+	pd.ents = make([]p9.Dirent, 0, len(pins))
+
+	// temporary conversion storage
+	attr := &p9.Attr{}
+	requestType := p9.AttrMask{Mode: true}
+
+	// actual conversion
+	for i, pin := range pins {
+		callCtx, cancel := pd.callCtx()
+		ipldNode, err := pd.core.ResolveNode(callCtx, pin.Path())
+		if err != nil {
+			cancel()
+			return qid, 0, err
+		}
+		if _, err = ipldStat(callCtx, attr, ipldNode, requestType); err != nil {
+			cancel()
+			return qid, 0, err
+		}
+
+		pd.ents = append(pd.ents, p9.Dirent{
+			Name:   gopath.Base(pin.Path().String()),
+			Offset: uint64(i + 1),
+			QID: p9.QID{
+				Type: attr.Mode.QIDType(),
+				Path: cidToQPath(ipldNode.Cid()),
+			},
+		})
+		cancel()
+	}
+
+	return qid, ipfsBlockSize, nil
+}
+
+func (pd *PinFS) Readdir(offset uint64, count uint32) ([]p9.Dirent, error) {
+	pd.Logger.Debugf("Readdir")
+
+	if pd.ents == nil {
+		return nil, fmt.Errorf("directory %q is not open for reading", pd.String())
+	}
+
+	return flatReaddir(pd.ents, offset, count)
+}
+
+func (pd *PinFS) Backtrack() (walkRef, error) {
+	// return the parent if we are the root
+	if len(pd.Trail) == 0 {
+		return pd.parent, nil
```

Comment:
We generalized this into the base class

https://github.com/ipfs/go-ipfs/pull/6612/commits/9ef8e6ad9c4fc3f0213cea2836e9baf585c7987e#diff-6c2c294fc60db12a68ce40457ff3ea5eR250
I'm not sure how acceptable the hacks around the type system are though.
Since the base method can't return a pointer to the superset object, we pass it in as `self` when called from the superset.
```go
func (id *IPFS) Backtrack() (fsutils.WalkRef, error) {
    return id.IPFSBase.backtrack(id)
}
```

We could make a procedure that's separate from the base.
(note this has side effects that will "move" the fid by modify its path if its `fid.Trail` is not empty, so it may be better to keep it as is)
``` go
func backtrack(pathTrail []string, self, parent WalkRef) (WalkRef, error) {...}

func (id *IPFS) Backtrack() (fsutils.WalkRef, error) {
    return backtrack(id.Trail, id, id.parent)
}
```

Thoughts?

Reply:
For now, instead of using the base class, each FS implements `..` handling independently.
https://github.com/djdv/go-ipfs/commit/0b1fe2019c3d43da7e9e0c5a7985a65847a85201#diff-06e72cc91b46c3fd9ec80eb3bc4c9388R334-R346
https://github.com/djdv/go-ipfs/commit/0b1fe2019c3d43da7e9e0c5a7985a65847a85201#diff-cbe47baa6ebcd4d1495ff0145fc9c10eR230-R235
etc...